File: searching-algorithms/binary_search.py
import math

def is_list_sorted(input_list):
    sorted_lst = True
    for idx, each_item in enumerate(input_list):
        if idx == 0:
            continue
        else:
            current_item = each_item
            previous_item = input_list[idx - 1]
            if current_item < previous_item:
                sorted_lst = False
                break

    return sorted_lst

def run_binary_search(search_number, input_list):

    if not is_list_sorted(input_list):
        raise Exception("List is not sorted in ascending order")
    
    number_index = -1
    start_index = 0 
    end_index = len(input_list)

    while start_index <= end_index:
        search_index = math.floor((start_index + end_index) / 2)
        print(search_index)

        if search_number == input_list[search_index]:
            number_index = search_index
            break
        
        elif search_number > input_list[search_index]:
            start_index = search_index + 1
            end_index = end_index
        
        else:
            start_index = 0
            end_index = search_index - 1

    return number_index


if __name__ == "__main__":
    search_number = 10
    input_list = [3, 4, 10, 11]
    number_index = run_binary_search(search_number, input_list)
    print("number found in index : ", number_index)

File: searching-algorithms/binary_search.js
"use strict";

// Note: Not being used currently
const isSortedArray = (inputArray) => {
  return new Promise((resolve, reject) => {
    for (let i = 0; i < inputArray.length; i++) {
      if (i == 0) {
        continue;
      }

      if (inputArray[i - 1] > inputArray[i]) {
        reject(false);
      }
    }

    resolve(true);
  });
};

const binarySearch = (searchNumber, inputArray) => {
  try {
    for (let i = 0; i < inputArray.length; i++) {
      if (i == 0) {
        continue;
      } else {
        if (inputArray[i - 1] > inputArray[i]) {
          throw "Array is not sorted in ascending order";
        }
      }
    }

    let numberIndex = -1;
    let startIndex = 0;
    let endIndex = inputArray.length;

    while (startIndex <= endIndex) {
      let searchIndex = Math.floor((startIndex + endIndex) / 2);

      if (searchNumber == inputArray[searchIndex]) {
        numberIndex = searchIndex;
        break;
      } else if (searchNumber > inputArray[searchIndex]) {
        startIndex = searchIndex + 1;
        endIndex = inputArray.length;
      } else {
        startIndex = 0;
        endIndex = searchIndex - 1;
      }
    }

    return numberIndex;
  } catch (err) {
    console.log("binarySearch error : ", err);
  }
};

let inputArray = [1, 3, 4, 11, 13];
let searchNumber = 13;

let numberIndex = binarySearch(searchNumber, inputArray);
console.log("numberIndex : ", numberIndex);
